libPhoneNumberGeocoding/NBPhoneNumberOfflineGeocoder.m

```diff
@@ -0,0 +1,150 @@
+//
+//  NBPhoneNumberOfflineGeocoder.m
+//  libPhoneNumberiOS
+//
+//  Created by Rastaar Haghi on 6/12/20.
+//  Copyright © 2020 Google LLC. All rights reserved.
+//
+
+#import "NBPhoneNumberOfflineGeocoder.h"
+#import "Metadata/NBGeocoderMetadataHelper.h"
+#import "NBPhoneNumber.h"
+#import "NBPhoneNumberUtil.h"
+
+@implementation NBPhoneNumberOfflineGeocoder {
+ @private
+  NBPhoneNumberUtil *_phoneNumberUtil;
+  NSCache<NSString *, NBGeocoderMetadataHelper *> *_metadataHelpers;
+}
+
+- (instancetype)init {
+  self = [super init];
+  if (self != nil) {
+    _phoneNumberUtil = NBPhoneNumberUtil.sharedInstance;
+    _metadataHelpers = [[NSCache alloc] init];
+  }
+  return self;
+}
+
+- (nullable NSString *)countryNameForNumber:(NBPhoneNumber *)number
+                           withLanguageCode:(NSString *)languageCode {
+  NSArray *regionCodes = [_phoneNumberUtil getRegionCodesForCountryCode:number.countryCode];
+  if ([regionCodes count] == 1) {
+    return [self regionDisplayName:regionCodes[0] withLanguageCode:languageCode];
+  } else {
+    NSString *regionWhereNumberIsValid = @"ZZ";
+    for (NSString *regionCode in regionCodes) {
+      if ([_phoneNumberUtil isValidNumberForRegion:number regionCode:regionCode]) {
+        if (![regionWhereNumberIsValid isEqualToString:@"ZZ"]) {
+          return nil;
+        }
+        regionWhereNumberIsValid = regionCode;
+      }
+    }
+
+    return [self regionDisplayName:regionWhereNumberIsValid withLanguageCode:languageCode];
+  }
+}
+
+- (nullable NSString *)regionDisplayName:(NSString *)regionCode
+                        withLanguageCode:(NSString *)languageCode {
+  if (regionCode == nil || [regionCode isEqualToString:@"ZZ"] ||
+      [regionCode isEqual:NB_REGION_CODE_FOR_NON_GEO_ENTITY]) {
+    return nil;
+  } else {
+    return [[NSLocale localeWithLocaleIdentifier:languageCode] displayNameForKey:NSLocaleCountryCode
+                                                                           value:regionCode];
+  }
+}
+
+- (nullable NSString *)descriptionForValidNumber:(NBPhoneNumber *)phoneNumber
+                                withLanguageCode:(NSString *)languageCode {
+  // If the NSCache doesn't contain a key equivalent to languageCode, create a new
+  // NBGeocoderMetadataHelper object with a language set equal to languageCode and
+  // default country code to United States / Canada
+  if ([_metadataHelpers objectForKey:languageCode] == nil) {
+    [_metadataHelpers setObject:[[NBGeocoderMetadataHelper alloc] initWithCountryCode:@1
```

Comment:
This one should be set to a constant like "US_REGION_CODE" or something.

Reply:
Done.